### ISSUE TITLE
request: aws_iam_policy_document pre-flight checks

### ISSUE THREAD

**Issue by hashibot** [hashicorp/terraform-provider-aws#491]
_This issue was originally opened by @therealbill as hashicorp/terraform#11444. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

# Affected Resource(s)
- aws_iam_policy_document

AWS has a hard limit on the size in characters of a policy document (the JSON it uses) of 2048 bytes. Trying to add a policy larger than that will result in an error on apply. 
# Current Behavior
The plan "works", the apply will fail.

# Desired Behavior
I would rather see the plan indicate the policy it is going to write is too large so it can be caught in that stage

In this case, terraform knows the size of the JSON document it just created and will attempt to upload, so I think it reasonable for it to check since it does not need info from the provider (this is one of the hard limits, you can't get them to raise it for you).

Also, the value of "sid" can only be alphanumeric, so I'd like to see the plan report an error if non-alphanumeric characters are in it. Less important then the policy size, of course.

**Comment by nitrocode** [comment]
This is planned in tflint https://github.com/wata727/tflint/issues/259

**Comment by taylorsmithgg** [comment]
As of terraform v0.12.13 and provider v2.35.0:
The validation error is coming back from AWS, but it is still creating the user with the policy, which then cannot be modified and has to be manually deleted and recreated.

```
Error: Error putting IAM user policy s3-fleetservices-prod-rwp: LimitExceeded: Maximum policy size of 2048 bytes exceeded for user s3-fleetservices-prod
    status code: 409, request id: c7f0f360-5d6f-47d7-9dc8-9ac1c2d2ab82

  on ../../../modules/s3_bucket_adjustable_perm/main.tf line 12, in resource "aws_iam_user_policy" "user_rwp_on_bucket":
  12: resource "aws_iam_user_policy" "user_rwp_on_bucket" {
```

**Comment by kunickiaj** [comment]
This seems like a good idea to me -- even if not implemented in the provider, TF 0.13 looks like it'll support user-supplied validation checks for resources that could work well for this use case.
